src/main/java/org/breedinginsight/brapps/importer/services/processors/experiment/create/workflow/CreateNewExperimentWorkflow.java

```diff
@@ -0,0 +1,308 @@
+        if (validationErrors.hasErrors()) {
+            throw new ValidatorException(validationErrors);
+        }
+
+        // TODO: move to experiment import service
+        ImportPreviewResponse response = buildImportPreviewResponse(importRows, processContext.getPendingData(), processedData, upload);
+
+        statusService.updateMappedData(upload, response, "Finished mapping data to brapi objects");
+
+        // preview data
+        if (!commit) {
+            statusService.updateOk(upload);
+            return response;
+        }
+
+        // commit data
+        long totalObjects = getNewObjectCount(response);
+        statusService.startUpload(upload, totalObjects, "Starting upload to brapi service");
+        statusService.updateMessage(upload, "Creating new experiment objects in brapi service");
+
+        commitPendingImportObjectsStep.process(processContext, processedData);
+
+        statusService.finishUpload(upload, totalObjects, "Completed upload to brapi service");
+        return response;
+    }
+
+    /**
+     * Process the import service context and returns an Optional ImportWorkflowResult.
+     *
+     * @param context The import service context to be processed. If null, then it skips processing but returns the result with no-preview.
+     * @return An Optional ImportWorkflowResult which contains the workflow and import preview response (if available).
+     *         If the context is null, it returns the result with no-preview.
+     * @throws Exception If any error occurs during the processing.
+     */
+    public Optional<ImportWorkflowResult> process(ImportServiceContext context) throws Exception {
+        // Workflow processing the context
+        ImportWorkflow workflow = ImportWorkflow.builder()
+                .id(getWorkflow().getId())
+                .name(getWorkflow().getName())
+                .build();
+
+        // No-preview result
+        Optional<ImportWorkflowResult> result;
+
+        result = Optional.of(ImportWorkflowResult.builder()
+                .workflow(workflow)
+                .importPreviewResponse(Optional.empty())
+                .build());
+
+        // Skip this workflow unless creating a new experiment
+        if (context != null && !this.workflow.isEqual(context.getWorkflowId())) {
+            return Optional.empty();
+        }
+
+        // Skip processing if no context, but return no-preview result for this workflow
+        if (context == null) {
+            return result;
+        }
+
+        // TODO: unify usage of single import context type throughout
+        ImportContext importContext = ImportContext.from(context);
+
+        // Start processing the import...
+        ImportPreviewResponse response = runWorkflow(importContext);
+
+        result = Optional.of(ImportWorkflowResult.builder()
+                .workflow(workflow)
+                .importPreviewResponse(Optional.of(response))
+                .build());
```

Comment:
When merging with BI-2134, just wrap the runWorkflow() call in a try-catch as below, and any exceptions will be passed and handled by the higher levels as usual. No other changes are needed.
```suggestion
        // Start processing the import...
        try {
            ImportPreviewResponse successResponse;
            successResponse = runWorkflow(importContext);
            result.ifPresent(importWorkflowResult ->   importWorkflowResult.setImportPreviewResponse(Optional.of(successResponse)));
        } catch (Exception e) {
            result.ifPresent(importWorkflowResult -> importWorkflowResult.setCaughtException(Optional.of(e)));
        }
       
       return result;
```

Reply:
Thanks, just about have it done. Need to remove Exception stuff on method signatures.

Reply:
Pushed changes